src/v/cloud_storage/remote_segment.cc

```diff
@@ -909,30 +915,17 @@ class remote_segment_batch_consumer : public storage::batch_consumer {
               "remote_segment_batch_consumer not initialized",
               _parent._seg->get_ntp());
 
-            if (
-              _parent._cur_ot_state->get().last_gap_offset()
-              < header.base_offset) {
-                _parent._cur_ot_state->get().add_gap(
-                  header.base_offset, header.last_offset());
-                vlog(
-                  _ctxlog.debug,
-                  "added offset translation gap [{}-{}], current state: {}",
-                  header.base_offset,
-                  header.last_offset(),
-                  _parent._cur_ot_state);
-
-                _parent._cur_delta += header.last_offset_delta
-                                      + model::offset{1};
-            } else {
-                // This can happen if we're dealing with overlapping segments
-                vlog(
-                  _ctxlog.debug,
-                  "offset translation gap [{}-{}] is already added, current "
-                  "state: {}",
-                  header.base_offset,
-                  header.last_offset(),
-                  _parent._cur_ot_state);
-            }
+            vlog(
+              _ctxlog.debug,
+              "added offset translation gap [{}-{}], current state: {}",
+              header.base_offset,
+              header.last_offset(),
+              _parent._cur_ot_state);
+
+            _parent._cur_ot_state->get().add_gap(
+              header.base_offset, header.last_offset());
+            _parent._cur_delta = model::offset_delta(
+              _parent._cur_ot_state->get().last_delta());
```

Comment:
But what if current offset is not the last offset that we've seen? Then it will diverge from the true delta.

I think it should just be unconditional `+= header.last_offset_delta + 1` because the segment reader always reads one segment in order without any duplicates.